In [2]:
import pandas as pd
import numpy as np

In [4]:
df = pd.read_excel('dados_vendas.xlsx')

In [37]:
df.head()

COD_FILIAL ORCAMENTO  DATA_COMPRA                 HORA COD_CLIENTE SEXO  \
0           1    99AWBR     20200515  2020-05-15T17:32:29    05090301    F   
1           1    99AW74     20200515  2020-05-15T11:14:54    05190001    M   
2          31    059559     20200515  2020-05-15T12:33:16    99DL9N01    F   
3          31    059584     20200515  2020-05-15T12:56:28    99DL9N01    F   
4           1    99AWCA     20200515  2020-05-15T20:45:45    05992701  NaN   

    NACIMENTO  IDADE  COD_PRODUTO  \
0  28/03/1964     56        25728   
1  12/11/1985     34        25172   
2  21/03/1986     34        26243   
3  21/03/1986     34        26243   
4  28/04/1954     66        26574   

                                    NOME_PRODUTO  \
0  SMARTPHONE SAMSUNG GAL J4 CORE 16GB PRETO       
1  SMARTPHONE LG K9 TV LM-X210BMW PRETO            
2  SUPORTE UNIV. BEDIN BU-25                       
3  SUPORTE UNIV. BEDIN BU-25                       
4  SMARTPHONE SAMSUNG GAL A30S 64GB BRANCO         

                                       CLASSIFICACAO  QUANTIDADE      CANAL  
0  CELULAR SMARTPHONE SMARTPHONE                 ...           1  ECOMMERCE  
1  CELULAR SMARTPHONE SMARTPHONE                 ...           1   VENDAWEB  
2  TECNOLOGIA TV E VIDEO SUPORTE PARA TV         ...           1  ECOMMERCE  
3  TECNOLOGIA TV E VIDEO SUPORTE PARA TV         ...           1  ECOMMERCE  
4  CELULAR SMARTPHONE SMARTPHONE                 ...           1  ECOMMERCE

In [5]:
df.COD_PRODUTO.value_counts()

26243    14001
21653    12909
25727    10397
25729    10264
25726     8090
         ...  
17196        1
19371        1
21160        1
18920        1
21362        1
Name: COD_PRODUTO, Length: 715, dtype: int64

In [ ]:
compras=df.COD_CLIENTE.value_counts().to_frame()
#compras[compras['COD_CLIENTE']>1].describe()
#compras.describe()

In [ ]:
n=len(compras)
slices=10
print("%\tCount\tMin\tmean\tstd")
for x in range(1,slices):
    percentage=(slices-x)*(100/slices)
    data=compras[:int(-n*(x/slices))]['COD_CLIENTE']
    print(str(percentage)+"%",data.count(),data.min(),"%.2f"%data.mean(),"%.2f"%data.std(),sep='\t')

In [136]:
vendas=df.COD_PRODUTO.value_counts().to_frame()
n=len(vendas)
slices=10
print("%\tCount\tMin\tmean\tstd")
for x in range(1,slices):
    percentage=(slices-x)*(100/slices)
    data=vendas[:int(-n*(x/slices))]['COD_PRODUTO']
    print(str(percentage)+"%",data.count(),data.min(),"%.2f"%data.mean(),"%.2f"%data.std(),sep='\t')

%	Count	Min	mean	std
90.0%	644	1	499.49	1348.54
80.0%	572	1	562.23	1418.66
70.0%	501	2	641.70	1499.13
60.0%	429	4	748.98	1595.36
50.0%	358	14	896.09	1708.87
40.0%	286	61	1113.49	1849.90
30.0%	215	198	1442.33	2029.60
20.0%	143	509	2001.41	2294.39
10.0%	72	1093	3236.38	2716.92


In [139]:
vendas=df.COD_PRODUTO.value_counts().to_frame()
print(vendas[:58])

       COD_PRODUTO
26243        14001
21653        12909
25727        10397
25729        10264
25726         8090
25981         7633
25996         6935
25982         6855
26222         6160
25983         5675
25354         5239
25728         5142
25172         4662
25481         4620
21647         4318
25545         4105
25303         4011
23802         3912
26005         3788
26004         3560
25676         3429
26003         3374
21255         3369
25173         3106
26324         3032
19729         3020
25355         2994
25998         2951
25546         2777
25997         2752
21854         2716
26247         2686
21256         2598
21855         2431
24913         2429
25870         2420
24716         2409
25366         2322
25933         2198
22752         2157
25730         2096
26423         1957
25577         1805
26325         1745
26007         1705
14537         1703
20966         1651
26387         1615
25904         1609
25675         1608
26008         1592
25193       

In [131]:
vendas=df.COD_PRODUTO.value_counts().to_frame()
first=np.percentile(vendas, 25)  
third=np.percentile(vendas, 75)
IQR=third-first
print("IQR: ",IQR)
outlier_threshold=3*IQR+third
n_antigo=len(vendas)
vendas=vendas[vendas['COD_PRODUTO']<outlier_threshold]
n=len(vendas)
print(n_antigo-n," Outliers excluidos")
slices=10
print("%\tCount\tMin\tmean\tstd")
for x in range(1,slices):
    percentage=(slices-x)*(100/slices)
    data=vendas[:int(-n*(x/slices))]['COD_PRODUTO']
    print(str(percentage)+"%",data.count(),data.min(),"%.2f"%data.mean(),"%.2f"%data.std(),sep='\t')

IQR:  321.0
58  Outliers excluidos
%	Count	Min	mean	std
90.0%	592	1	177.45	297.60
80.0%	526	1	199.59	308.70
70.0%	460	2	228.05	320.20
60.0%	395	3	265.22	331.12
50.0%	329	8	317.51	339.54
40.0%	263	33	393.00	340.30
30.0%	198	102	502.62	324.06
20.0%	132	286	662.33	281.71
10.0%	66	607	896.53	197.75


In [24]:
for x in range(10):
    teste= pd.qcut(df.COD_PRODUTO,25)
    break
teste

0         (25727.0, 25729.0]
1         (24851.0, 25189.0]
2         (26178.0, 26243.0]
3         (26178.0, 26243.0]
4         (26572.0, 26727.0]
                 ...        
321736    (23802.0, 24851.0]
321737    (25869.0, 25981.0]
321738    (25869.0, 25981.0]
321739    (25718.0, 25727.0]
321740    (25718.0, 25727.0]
Name: COD_PRODUTO, Length: 321741, dtype: category
Categories (25, interval[float64]): [(10215.999, 21029.0] < (21029.0, 21653.0] < (21653.0, 21821.0] < (21821.0, 22617.0] ... (26243.0, 26324.0] < (26324.0, 26424.0] < (26424.0, 26572.0] < (26572.0, 26727.0]]

Número de clientes = 236244 <br>
Número de produtos = 715  <br>
Número de compras  = 321741 <br>
Número de compras sem repetição = 311741 <br>
Número de compradores com mais de uma compra= 57365 <br>

In [6]:
newdf= df[['COD_CLIENTE','COD_PRODUTO']]
newdf=newdf.drop_duplicates(keep='first')
print(newdf.shape)
v=newdf['COD_CLIENTE'].nunique()
u=newdf['COD_PRODUTO'].nunique()

pd.set_option('display.max_rows', 500)

convert_cliente={}
revert_cliente=['' for x in range(v)]
c=0
for x in newdf['COD_CLIENTE'].unique():
    convert_cliente[x]=c
    revert_cliente[c]=x
    c+=1
convert_produto={}
revert_produto=['' for x in range(u)]
c=0
for x in newdf['COD_PRODUTO'].unique():
    convert_produto[x]=c
    revert_produto[c]=x
    c+=1
lista_adj=[[] for x in range(v)]
lista_adj_u=[[] for x in range(u)]
for i,j in newdf.iterrows():
    lista_adj[convert_cliente[j[0]]].append(convert_produto[j[1]])
    lista_adj_u[convert_produto[j[1]]].append(convert_cliente[j[0]])
#teste=[x for l in lista_adj for x in l]
#pd.set_option('display.max_rows', 500)
#produtos_tratado=pd.DataFrame.from_dict({x:teste.count(x) for x in teste}, orient='index').sort_values(0,ascending=False)
#print(produtos_tratado.describe())
outputname="Adjacency_list.txt"
text_file = open(outputname, "w")
text_file.write(str(v)+'\n')
text_file.write(str(u)+'\n')
for j in lista_adj:
    txt=str(j)[1:-1].replace(',',' ')
    text_file.write( txt+'\n')
text_file.close()

#p = Popen(['./AMRPB',outputname], shell=True, stdout=PIPE, stdin=PIPE)


(311741, 2)


In [7]:
f = open("saida.txt", "r")
nonnumbers=['(',')','}',']']
text=f.read().split('}')
clustertext=text[0].split('[')
solotext=text[1].split('(')
#print(len(solotext[2]))
biclusters=[]
v_cluster=[[] for x in range(v)]
u_cluster=[[] for x in range(u)]
#Clusters
for i,a in enumerate(clustertext[1:]):
    bicluster=[]
    for j,b in enumerate(a.split('),')):
        array=[]
        for k,c in enumerate(b.replace('],','').split(',')):
            txt=c
            for r in nonnumbers:
                txt=txt.replace(r, "")
            if(txt!=''):
                code=int(txt)
                array.append(code)
                if(not j):
                    v_cluster[code].append(i)
                else:
                    u_cluster[code].append(i)
        bicluster.append(array)
    biclusters.append(bicluster)
#Solos
#for i,a in enumerate(solotext[1:]):
#    for j,b in enumerate(a.split(',')):
#        txt=b
#        for r in nonnumbers:
#            txt=txt.replace(r, "")
#       if(txt!=''):
#           code=int(txt)
#            new_row = {'code':code, 'cluster':None }
#            if(j):
#               v_cluster[code].append()
#           else:
#                u_cluster[code].append()
print(u_cluster)

[[144, 230, 239, 267], [140, 141, 198, 207, 230, 245, 246], [25, 31, 49, 55, 78, 87, 115, 117, 118, 124, 144, 175, 178, 201, 219, 231, 233, 236, 240, 244, 257, 258, 261, 267], [], [146], [224, 245], [], [], [196], [44, 280, 295], [], [137, 199, 230, 240, 248], [], [230], [85, 137, 252, 283], [196], [4, 11, 15, 21, 24, 26, 27, 29, 30, 47, 56, 68, 80, 93, 108, 109, 117, 128, 129, 130, 132, 157, 163, 168, 174, 176, 181, 186, 188, 192, 201, 216, 230, 234, 243, 245, 248, 267, 276, 281, 283], [], [], [25, 175], [0, 63], [230], [], [], [207, 211, 219], [], [141], [], [141], [148], [], [145, 196, 198, 202, 225, 248, 260, 272], [211], [], [], [], [], [], [], [139, 148, 239, 240, 257, 272], [271], [230], [], [], [], [], [74, 139], [207, 225, 230, 250], [230], [], [237], [], [123, 145], [140, 230], [], [18, 182], [201], [], [], [95, 135, 139, 207, 230, 253, 268], [], [33, 195], [207], [110], [], [], [247], [], [], [], [139], [], [111], [276], [], [53], [], [66], [267], [], [], [], [5, 171], [230]

In [ ]:
print(v_cluster)

# Get recomendations

### Calcula similaridade

In [34]:
def similarity (cod_viewed,cod_recomended,lista_adj_u,cluster_u):
    total=len(cluster_u)
    partial=len(set(cluster_u) & set(lista_adj_u[cod_recomended]))
    #print(total,partial)
    #print(set(recomendations) & set(lista_adj_u[cod_recomended]),sep='\n')
    if(partial==0):
        return 0
    return partial/total
        

In [32]:
#def similarity (cod_viewed,cod_recomended,lista_adj_u,recomendations):
#    total=len(recomendations)
#    clusters
#    partial=len(set(recomendations) & set(lista_adj_u[cod_recomended]))
#    print(total,partial)
#    print(set(recomendations) & set(lista_adj_u[cod_recomended]),sep='\n')
#    if(partial==0):
#        return 0
#    return partial/total
        

### Cliente

In [31]:
cliente='9NW44N01'
try:
    recomendations=[]
    cliente_key=convert_cliente[cliente]
    clusters=v_cluster[cliente_key]
    for a in clusters:
        for j,b in enumerate(biclusters[a]):
            if(j==1):
                for x in b:
                    recomendations.append(x)
    print(recomendations)
except KeyError:
    print("Cliente não encontrado")
except:
    print("Erro inesperado encontrado, contacte o administrador")


[323]


### Produto

In [36]:
produto=26243
#try:
recomendations=[]
produto_key=convert_produto[produto]
clusters=u_cluster[produto_key]
scores=[]
print(clusters)
for a in clusters:
    for j,b in enumerate(biclusters[a]):
        if(j==1):
            for x in b:
                if(x not in recomendations and x!= produto_key):
                    recomendations.append(x)
                    scores.append(similarity(produto_key,x,lista_adj_u,b))
#print(recomendations)

#print("SCORES")
print(scores)
#except KeyError:
#    print("Cliente não encontrado")
#except:
#    print("Erro inesperado encontrado, contacte o administrador")


[25, 31, 49, 55, 78, 87, 115, 117, 118, 124, 144, 175, 178, 201, 219, 231, 233, 236, 240, 244, 257, 258, 261, 267]
[0, 0, 0, 0, 0, 0, 0, 0, 0.3333333333333333, 0, 0, 0, 0.2857142857142857, 0, 0, 0, 0, 0, 0, 0, 0, 0.3333333333333333, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.125, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


### Usando vizinhos segundos para clientes

In [37]:
cliente='9NW44N01'
try:
    recomendations=[]
    cliente_key=convert_cliente[cliente]
    clusters=v_cluster[cliente_key]
    for a in clusters:
        for j,b in enumerate(biclusters[a]):
            if(j==1):
                for x in b:
                    recomendations.append(x)
            else:
                for viz in b:
                    for clu in v_cluster[viz]:
                        for i,el in enumerate(biclusters[clu]):
                            if(i==1):
                                for pro in el:
                                    if(pro not in recomendations):
                                        print(pro)
                                        recomendations.append(pro)
    comprados=[]
    compras_cliente=newdf[newdf.COD_CLIENTE==cliente]
    for x in compras_cliente.COD_PRODUTO:
        comprados.append(convert_produto[x])
    recomendations=list(filter(lambda a: a not in comprados, recomendations))
    print(recomendations)
except KeyError:
    print("Cliente não encontrado")
except:
    print("Erro inesperado encontrado, contacte o administrador")


323
272
[272]


### Usando vizinhos segundos para produtos

In [30]:
produto=26243
try:
    recomendations=[]
    produto_key=convert_produto[produto]
    clusters=u_cluster[cliente_key]
    for a in clusters:
        for j,b in enumerate(biclusters[a]):
            if(j==0):
                for x in b:
                    recomendations.append(x)
            else:
                for viz in b:
                    for clu in v_cluster[viz]:
                        for i,el in enumerate(biclusters[clu]):
                            if(i==1):
                                for pro in el:
                                    if(pro not in recomendations):
                                        print(pro)
                                        recomendations.append(pro)
    print(recomendations)
except KeyError:
    print("Cliente não encontrado")
except:
    print("Erro inesperado encontrado, contacte o administrador")


323
20
[16347, 54748, 90256, 323, 3481, 41518, 20]


In [ ]:
import shlex
from subprocess import Popen, PIPE
def get_exitcode_stdout_stderr(cmd):
    """
    Execute the external command and get its exitcode, stdout and stderr.
    """
    args = shlex.split(cmd)

    proc = Popen(args, stdout=PIPE, stderr=PIPE)
    out, err = proc.communicate()
    exitcode = proc.returncode
    #
    return exitcode, out, err

cmd = './AMRPB Adjacency_list.txt'  # arbitrary external command, e.g. "python mytest.py"
exitcode, out, err = get_exitcode_stdout_stderr(cmd)

In [4]:
f = out.decode("utf-8").split('\n')[-1]
nonnumbers=['(',')','}',']']
text=f.split('}')
clustertext=text[0].split('[')
solotext=text[1].split('(')
#print(len(solotext[2]))
biclusters=[]
v_cluster=[[] for x in range(v)]
u_cluster=[[] for x in range(u)]
#Clusters
for i,a in enumerate(clustertext[1:]):
    bicluster=[]
    for j,b in enumerate(a.split('),')):
        array=[]
        for k,c in enumerate(b.replace('],','').split(',')):
            txt=c
            for r in nonnumbers:
                txt=txt.replace(r, "")
            if(txt!=''):
                code=int(txt)
                array.append(code)
                if(not j):
                    v_cluster[code].append(i)
                else:
                    u_cluster[code].append(i)
        bicluster.append(array)
    biclusters.append(bicluster)
#Solos
#for i,a in enumerate(solotext[1:]):
#    for j,b in enumerate(a.split(',')):
#        txt=b
#        for r in nonnumbers:
#            txt=txt.replace(r, "")
#       if(txt!=''):
#           code=int(txt)
#            new_row = {'code':code, 'cluster':None }
#            if(j):
#               v_cluster[code].append()
#           else:
#                u_cluster[code].append()
print(u_cluster)

NameError: name 'v' is not defined